src/InteractiveWindow/Editor/InteractiveWindow.cs

```diff
@@ -1447,88 +1461,126 @@ private void MeasurePrompts(int startLine, int endLine, out int minPromptLength,
+            var kind = GetSpanKind(snapshot);
+            return (kind == ReplSpanKind.Prompt) || (kind == ReplSpanKind.SecondaryPrompt);
         }
 
-        private class ReplSpanComparer : IComparer<ReplSpan>
+        private bool IsLanguage(ITextSnapshot snapshot)
         {
-            public static readonly IComparer<ReplSpan> Instance = new ReplSpanComparer();
+            return GetSpanKind(snapshot) == ReplSpanKind.Language;
+        }
 
-            private ReplSpanComparer()
-            {
-            }
+        private static ReadOnlyCollection<SnapshotSpan> GetSourceSpans(ITextSnapshot snapshot)
+        {
+            return ((IProjectionSnapshot)snapshot).GetSourceSpans();
+        }
 
-            int IComparer<ReplSpan>.Compare(ReplSpan x, ReplSpan y)
+        private int GetPromptIndexForPoint(ReadOnlyCollection<SnapshotSpan> sourceSpans, SnapshotPoint point)
+        {
+            int index = BinarySearch(
+                sourceSpans,
+                sourceSpan =>
+                {
+                    var start = _textView.BufferGraph.MapUpToBuffer(sourceSpan.Start, PointTrackingMode.Positive, PositionAffinity.Successor, _projectionBuffer);
+                    Debug.Assert(start != null);
+                    if (point < start.Value)
+                    {
+                        return -1;
+                    }
+                    var end = _textView.BufferGraph.MapUpToBuffer(sourceSpan.End, PointTrackingMode.Positive, PositionAffinity.Successor, _projectionBuffer);
+                    Debug.Assert(end != null);
+                    return (point < end.Value) ? 0 : 1;
+                });
+
+            if (index < 0)
             {
-                int comp = x.LineNumber - y.LineNumber;
-                return comp != 0
-                    ? comp
-                    : GetKindOrdinal(x) - GetKindOrdinal(y);
+                Debug.Assert(~index == sourceSpans.Count);
+                Debug.Assert(point.Position == point.Snapshot.Length);
+                index = ~index - 1;
             }
 
-            /// <summary>
-            /// Within a given line, the order is Output &lt; Prompt &lt; Language/StdIn.
-            /// </summary>
-            /// <remarks>
-            /// While, visually, it appears that every line begins with a prompt, a line
-            /// may actually begin with an empty output prompt, because at any given time
-            /// there must be a writable output buffer for <see cref="AppendOutput"/>.
-            /// Generally, this occurs when a submission has no output.
-            /// </remarks>
-            private static int GetKindOrdinal(ReplSpan span)
+            Debug.Assert(index >= 0);
+            Debug.Assert(index < sourceSpans.Count);
+
+            // Find the nearest preceding prompt.
+            while (!IsPrompt(sourceSpans[index].Snapshot))
```

Comment:
It's okay to finish with a linear search because we expect the number of iterations to be small?